### PR TITLE
WIP: Add conformal predictions for regression

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Example code:

```python3
import pandas as pd
from sklearn.metrics import mean_squared_error

from fortuna.conformal import CVPlusConformalRegressor
from fortuna.metric.regression import prediction_interval_coverage_probability

from autogluon.tabular import TabularPredictor


# TODO: Make repeated bagging work
# TODO: Test for non-uniform indices
# TODO: Figure out if WeightedEnsemble can work
if __name__ == '__main__':
    label = 'age'
    train_data = pd.read_csv('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')
    subsample_size = 3000  # subsample subset of data for faster demo, try setting this to much larger values
    if subsample_size is not None and subsample_size < len(train_data):
        train_data = train_data.sample(n=subsample_size, random_state=0)
    train_data = train_data.reset_index(drop=True)  # FIXME: Errors if this isn't done
    test_data = pd.read_csv('https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')

    error = 0.2

    predictor: TabularPredictor = TabularPredictor(label=label, eval_metric='mse', problem_type='regression').fit(
        train_data, num_bag_folds=5,
        hyperparameters={'FASTAI': {}, 'GBM': {}, 'XGB': {}},
    )

    model_names = predictor.get_model_names()

    for model_name in model_names:
        model = predictor._trainer.load_model(model_name)
        num_folds = len(model.models)

        print(f'Conformal Predictions for {model_name}:')

        X_test_internal = predictor.transform_features(data=test_data, model=model_name)
        y_test_internal = predictor.transform_labels(labels=test_data[label])
        y_pred_test = model.predict(X_test_internal)

        if num_folds <= 1:
            print(f'\tSkipping due to num_folds={num_folds}')
            mse = mean_squared_error(y_true=y_test_internal, y_pred=y_pred_test)
            print(f'\t\tmse BagEns: {mse}')
            continue
        cross_val_outputs, cross_val_targets, cross_test_outputs = predictor.get_conformal_predictions(
            test_data=test_data,
            model=model_name,
        )
        n_splits = len(cross_test_outputs)
        y_pred_test_per_fold = model.predict_folds(X_test_internal)

        cvplus_interval = CVPlusConformalRegressor().conformal_interval(
            cross_val_outputs=cross_val_outputs,
            cross_val_targets=cross_val_targets,
            cross_test_outputs=cross_test_outputs,
            error=error,
        )
        cvplus_coverage = prediction_interval_coverage_probability(
            cvplus_interval[:, 0], cvplus_interval[:, 1], y_test_internal
        )
        mse = mean_squared_error(y_true=y_test_internal, y_pred=y_pred_test)
        print(f"\tDesired coverage: {1 - error:.4f}\n"
              f"\tCV+ empirical coverage: {cvplus_coverage:.4f}")
        for i in range(20):
            ground_truth = y_test_internal.iloc[i]
            interval = cvplus_interval[i, :]
            lower_interval, upper_interval = interval
            is_within_interval = (ground_truth >= lower_interval) and (ground_truth < upper_interval)
            print(f'\tGT: {ground_truth}\t| {lower_interval:.2f}\t| {upper_interval:.2f}\t| {is_within_interval}')
        print(f'\t\tmse BagEns: {mse:.2f}')
        for i in range(n_splits):
            mse_fold = mean_squared_error(y_true=y_test_internal, y_pred=cross_test_outputs[i])
            print(f'\t\tmse fold {i}: {mse_fold:.2f}')

```

Example output:
```
Conformal Predictions for LightGBM_BAG_L1:
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
/home/ubuntu/.conda/envs/code/lib/python3.8/site-packages/fortuna/metric/regression.py:128: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  targets = targets[:, None]
	Desired coverage: 0.8000
	CV+ empirical coverage: 0.8008
	GT: 31	| 28.44	| 55.29	| True
	GT: 17	| 13.60	| 39.06	| True
	GT: 47	| 31.70	| 57.04	| True
	GT: 21	| 10.50	| 36.01	| True
	GT: 17	| 9.17	| 34.43	| True
	GT: 58	| 31.87	| 57.16	| False
	GT: 42	| 30.20	| 55.90	| True
	GT: 60	| 33.62	| 58.86	| False
	GT: 20	| 25.91	| 51.51	| False
	GT: 29	| 26.48	| 52.05	| True
	GT: 36	| 32.84	| 58.73	| True
	GT: 72	| 40.90	| 66.44	| False
	GT: 37	| 22.39	| 47.75	| True
	GT: 43	| 33.36	| 58.88	| True
	GT: 67	| 44.75	| 70.54	| True
	GT: 40	| 30.58	| 55.85	| True
	GT: 49	| 45.04	| 70.76	| True
	GT: 52	| 33.81	| 59.15	| True
	GT: 45	| 34.87	| 60.26	| True
	GT: 19	| 7.71	| 33.16	| True
		mse BagEns: 105.70
		mse fold 0: 108.32
		mse fold 1: 108.35
		mse fold 2: 108.06
		mse fold 3: 106.65
		mse fold 4: 106.95
Conformal Predictions for NeuralNetFastAI_BAG_L1:
	Desired coverage: 0.8000
	CV+ empirical coverage: 0.8162
	GT: 31	| 21.00	| 46.90	| True
	GT: 17	| 12.20	| 37.74	| True
	GT: 47	| 29.79	| 55.46	| True
	GT: 21	| 11.07	| 36.47	| True
	GT: 17	| 6.28	| 31.95	| True
	GT: 58	| 32.56	| 57.97	| False
	GT: 42	| 31.93	| 57.25	| True
	GT: 60	| 31.15	| 56.72	| False
	GT: 20	| 20.44	| 45.78	| False
	GT: 29	| 24.69	| 50.73	| True
	GT: 36	| 36.06	| 61.49	| False
	GT: 72	| 44.26	| 69.51	| False
	GT: 37	| 19.95	| 45.70	| True
	GT: 43	| 33.78	| 60.25	| True
	GT: 67	| 39.44	| 65.74	| False
	GT: 40	| 29.92	| 55.32	| True
	GT: 49	| 31.10	| 58.06	| True
	GT: 52	| 31.01	| 56.36	| True
	GT: 45	| 36.91	| 62.79	| True
	GT: 19	| 7.89	| 34.14	| True
		mse BagEns: 103.81
		mse fold 0: 107.65
		mse fold 1: 108.11
		mse fold 2: 107.91
		mse fold 3: 106.78
		mse fold 4: 108.01
Conformal Predictions for XGBoost_BAG_L1:
/home/ubuntu/.conda/envs/code/lib/python3.8/site-packages/fortuna/metric/regression.py:128: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  targets = targets[:, None]
	Desired coverage: 0.8000
	CV+ empirical coverage: 0.8133
	GT: 31	| 27.03	| 53.80	| True
	GT: 17	| 12.27	| 38.07	| True
	GT: 47	| 30.06	| 56.11	| True
	GT: 21	| 10.04	| 35.85	| True
	GT: 17	| 6.50	| 32.18	| True
	GT: 58	| 30.74	| 56.49	| False
	GT: 42	| 30.03	| 55.85	| True
	GT: 60	| 32.07	| 57.74	| False
	GT: 20	| 22.02	| 48.37	| False
	GT: 29	| 27.74	| 53.39	| True
	GT: 36	| 31.35	| 57.33	| True
	GT: 72	| 48.51	| 75.71	| True
	GT: 37	| 25.46	| 51.25	| True
	GT: 43	| 30.24	| 55.96	| True
	GT: 67	| 42.30	| 68.55	| True
	GT: 40	| 27.07	| 52.84	| True
/home/ubuntu/.conda/envs/code/lib/python3.8/site-packages/fortuna/metric/regression.py:128: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  targets = targets[:, None]
	GT: 49	| 38.73	| 66.48	| True
	GT: 52	| 32.59	| 58.38	| True
	GT: 45	| 34.35	| 60.23	| True
	GT: 19	| 7.03	| 32.77	| True
		mse BagEns: 105.30
		mse fold 0: 109.28
		mse fold 1: 107.37
		mse fold 2: 107.26
		mse fold 3: 107.46
		mse fold 4: 107.87
Conformal Predictions for WeightedEnsemble_L2:
	Skipping due to num_folds=1
		mse BagEns: 102.36561723277092
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 1]
```diff
@@ -4138,6 +4138,32 @@ def _assert_is_fit(self, message_suffix: str = None):
                 error_message = f"{error_message} `.{message_suffix}`."
             raise AssertionError(error_message)
 
+    def get_conformal_predictions(self, test_data, model: str) -> (List[np.ndarray], List[np.ndarray], List[np.ndarray]):
```
Comment (gianlucadetommaso) [inline tabular/src/autogluon/tabular/predictor/predictor.py:4141]: The name here might be misleading, as this method does not strictly returns "conformal predictions", but it returns ensembles of outputs and targets used in the conformal method.

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 2]
```diff
@@ -356,6 +357,32 @@ def predict_proba(self, X, normalize=None, **kwargs):
...
 
+    def predict_folds(self, X, **kwargs):
+        y_pred_probas = self.predict_proba_folds(X, **kwargs)
+        y_preds = []
+        for y_pred_proba in y_pred_probas:
+            y_pred = get_pred_from_proba(y_pred_proba=y_pred_proba, problem_type=self.problem_type)
+            y_preds.append(y_pred)
+        return y_preds
+
+    def predict_proba_folds(self, X, normalize=None, **kwargs):
```
Comment (gianlucadetommaso) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:368]: Is this method available also in classification or only in regression? If it is available in regression, what would it return?

File: core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py [hunk 3]
```diff
@@ -356,6 +357,32 @@ def predict_proba(self, X, normalize=None, **kwargs):
...
+        for i, model in enumerate(self.models):
+            model = self.load_child(model)
+            if not has_preprocessed:
+                X = self.preprocess(X, model=model, **kwargs)
+                has_preprocessed = True
+            pred_proba = model.predict_proba(X=X, preprocess_nonadaptive=False, normalize=normalize)
+
+            if self.params_aux.get("temperature_scalar", None) is not None:
+                pred_proba = self._apply_temperature_scaling(pred_proba)
+            elif self.conformalize is not None:
```
Comment (gianlucadetommaso) [inline core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:380]: Should we allow for _both_ temperature scaling and conformal predictions?

An important note is that most conformal methods will not return calibrated probabilities (like temperature scaling would do), but they will rather return sets of likely predictions. In regression, these sets are usually expressed as intervals around the model prediction. In classification, these are discrete sets of likely variables (e.g. [1, 4, 6], for instance).

A few specific conformal method might be able to return calibrated probabilities, but it is not the standard output of conformal prediction. What output should we aim for?

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 4]
```diff
@@ -4138,6 +4138,32 @@ def _assert_is_fit(self, message_suffix: str = None):
...
+        self._assert_is_fit("get_conformal_predictions")
+        if self.problem_type != REGRESSION:
```
Comment (gianlucadetommaso) [inline tabular/src/autogluon/tabular/predictor/predictor.py:4143]: I assume `self.problem_type` can be `REGRESSION` or `CLASSIFICATION`? Do you have a separate case for binary classification, or is this part of `CLASSIFICATION`?

**Comment by gianlucadetommaso** [comment]
Please find below an example for binary classification. We use two conformal prediction methods from Fortuna: `SimplePredictionConformalClassifier` and `Multicalibrator`. The first methods returns conformal sets of likely labels at a desired coverage. The second method return calibrated probabilities that the target binary variable is 1. 

To evaluate the quality of the results, we use the empirical coverage in the first case, and the Mean Squared Error in the second.

```
import pandas as pd
from sklearn.metrics import accuracy_score
from fortuna.conformal import SimplePredictionConformalClassifier, BinaryClassificationMulticalibrator
import numpy as np
from autogluon.tabular import TabularPredictor

TRAIN_SIZE = 5000
VAL_SIZE = 5000
ERROR = 0.2
SUBSAMPLE_SIZE = 10000

if __name__ == '__main__':
    label = 'class'
    train_data = pd.read_csv('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')

    # subsample subset of data for faster demo, try setting this to much larger values
    # if SUBSAMPLE_SIZE is not None and SUBSAMPLE_SIZE < len(train_data):
    #     train_data = train_data.sample(n=SUBSAMPLE_SIZE, random_state=0)
    # train_data, val_data = train_data.iloc[:TRAIN_SIZE], train_data.iloc[TRAIN_SIZE:TRAIN_SIZE+VAL_SIZE]
    # train_data = train_data.reset_index(drop=True)  # FIXME: Errors if this isn't don
    train_data, val_data = train_data.iloc[:TRAIN_SIZE], train_data.iloc[TRAIN_SIZE:TRAIN_SIZE + VAL_SIZE]
    test_data = pd.read_csv('https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')

    predictor: TabularPredictor = TabularPredictor(label=label, eval_metric='accuracy', problem_type='multiclass').fit(
        train_data, num_bag_folds=5,
        hyperparameters={'FASTAI': {}, 'GBM': {}, 'XGB': {}},
    )

    model_names = predictor.get_model_names()

    for model_name in model_names:
        model = predictor._trainer.load_model(model_name)
        num_folds = len(model.models)

        print(f'Conformal Predictions for {model_name}:')

        X_val_internal = predictor.transform_features(data=val_data, model=model_name)
        y_val_internal = predictor.transform_labels(labels=val_data[label]).astype(int)
        y_pred_val = model.predict(X_val_internal)
        y_probs_val = model.predict_proba(X_val_internal)
        y_probs_val = np.stack([1 - y_probs_val, y_probs_val], axis=1)

        X_test_internal = predictor.transform_features(data=test_data, model=model_name)
        y_test_internal = predictor.transform_labels(labels=test_data[label]).astype(int)
        y_pred_test = model.predict(X_test_internal)
        y_probs_test = model.predict_proba(X_test_internal)
        y_probs_test = np.stack([1 - y_probs_test, y_probs_test], axis=1)

        acc = accuracy_score(y_true=y_test_internal, y_pred=y_pred_test)
        print(f'\tAccuracy: {acc:.2f}')

        conformal_sets = SimplePredictionConformalClassifier().conformal_set(
            val_probs=y_probs_val,
            val_targets=y_val_internal.values.astype(int),
            test_probs=y_probs_test,
            error=ERROR
        )

        empirical_coverage = np.mean([y in s for y, s in zip(y_test_internal, conformal_sets)])

        print(f"\nSimple Prediction Sets")
        print(f"\tDesired coverage: {1 - ERROR:.4f}\n"
              f"\tEmpirical coverage: {empirical_coverage:.4f}")

        mc = BinaryClassificationMulticalibrator()
        y_calib_probs_test, calib_status = mc.calibrate(
            targets=y_val_internal.values.astype(int),
            probs=y_probs_val[:, 1],
            test_probs=y_probs_test[:, 1],
        )

        print(f"\nBinary Classification Multicalibrator (no groups)")
        print(f"\tMSE before calibration: {mc.mean_squared_error(probs=y_probs_test[:, 1], targets=y_test_internal.values):.4f}")
        print(f"\tMSE after calibration: {mc.mean_squared_error(probs=y_calib_probs_test, targets=y_test_internal.values):.4f}")

        print(f"\tAccuracy before calibration: {accuracy_score(y_true=y_test_internal.values, y_pred=y_pred_test):.4f}")
        print(f"\tAccuracy after calibration: {accuracy_score(y_true=y_test_internal.values, y_pred=y_calib_probs_test > 0.5):.4f}")

```

Example output:
```
Conformal Predictions for LightGBM_BAG_L1:
	Accuracy: 0.86

	Simple Prediction Sets
	Desired coverage: 0.8000
	Empirical coverage: 0.8073

	Binary Classification Multicalibrator (no groups)
	MSE before calibration: 0.0950
	MSE after calibration: 0.0979
	Accuracy before calibration: 0.8644
	Accuracy after calibration: 0.8613
Conformal Predictions for NeuralNetFastAI_BAG_L1:
	Accuracy: 0.85

	Simple Prediction Sets
	Desired coverage: 0.8000
	Empirical coverage: 0.8027

	Binary Classification Multicalibrator (no groups)
	MSE before calibration: 0.0999
	MSE after calibration: 0.1023
	Accuracy before calibration: 0.8534
	Accuracy after calibration: 0.8535
Conformal Predictions for XGBoost_BAG_L1:
	Accuracy: 0.86

	Simple Prediction Sets
	Desired coverage: 0.8000
	Empirical coverage: 0.8025

	Binary Classification Multicalibrator (no groups)
	MSE before calibration: 0.0932
	MSE after calibration: 0.0978
	Accuracy before calibration: 0.8649
	Accuracy after calibration: 0.8597
Conformal Predictions for WeightedEnsemble_L2:
	Accuracy: 0.86

	Simple Prediction Sets
	Desired coverage: 0.8000
	Empirical coverage: 0.8025

	Binary Classification Multicalibrator (no groups)
	MSE before calibration: 0.0932
	MSE after calibration: 0.0978
	Accuracy before calibration: 0.8649
	Accuracy after calibration: 0.8597
```
